astacus/coordinator/plugins/base.py

```diff
@@ -66,7 +66,8 @@ def get_result(self, step_class: Type[Step[T]]) -> T:
 
     def set_result(self, step_class: Type[Step[T]], result: T) -> None:
         if step_class in self.step_results:
-            raise RuntimeError(f"result already set for step {step_class}")
+            if self.step_results[step_class] is not None or result is not None:
+                raise RuntimeError(f"result already set for step {step_class}")
```

Comment:
If I understand correctly, this is linked to `CassandraSubOpStep`.
I think it would be better to be a bit more verbose and create one short subclass for each op
instead of introducing an exception to the general behavior.

This will also make the progress logging more clear, since it relies on class name.

Reply:
I tried it first, but it looks pretty ugly. We're also doing CassandraStart twice, and some other things.

Reply:
Having ..Again classes looks pretty lame. If progress reporting relies on this, I'd rather just use index in the steps for that primarily.

Reply:
It's more than reporting.

I think keeping a zero-exception "each step has its class" rule has value:
 - rules without exception are easier to explain and reason about
 - communication between people during failure diagnosis is more straightforward if the-class-is-the-step ("the second StartCassandraStep got stuck, I mean the step 8, starting at zero, except after astacus 0.7 where it's step 9" vs "RestartCassandraStep got stuck").

Reply:
(or use the actual step instance and not its class name as key if we do not care about order - but I'm making order available for now)

Reply:
Currently the logic where each class is essentially singleton is quite counterintuitive. The fact they're used as nice hack to get previous steps' results is mainly awkward - it would be probably preferable to have e.g. per job (series of steps) state, which is typed, and populated by the steps as they come by the state.

Insofar reporting is concerned, if log says step M of N, quoting that M is sufficient to identify step. The class names should be implementation artifact.

Reply:
Possibly also refactoring if so that you can get result by type of result would make more sense as those ( if any ) seem to be unique.